R/add_cols.R

```diff
@@ -0,0 +1,247 @@
+#' Add event date as column to infectious history `<data.frame>`
+#'
+#' @description The event date could be first contact, last contact or other.
+#'
+#' @param .data A `<data.frame>` containing the infectious history from a
+#' branching process simulation.
+#' @param contact_type A `character` with the type of contact, either first
+#' contact (`"first"`), or last contact (`"last"`).
+#' @param distribution A `character` with the name of the distribution,
+#' following the base R convention for distribution naming (e.g. Poisson
+#' is `pois`).
+#' @param ... [dots] Extra arguments to be passed to the distribution function
+#' given in the `distribution` argument.
+#' @inheritParams sim_linelist
+#'
+#' @name .add_date
+#'
+#' @return A `<data.frame>` with one more column than input into `.data`.
+#' Unless the column heading is already present in which the data is
+#' overwritten.
+#' @keywords internal
+#'
+NULL
+
+#' @name .add_date
+.add_date_contact <- function(.data,
+                              contact_type = c("first", "last"),
+                              distribution = c("pois", "geom"),
+                              ...,
+                              outbreak_start_date) {
+  contact_type <- match.arg(contact_type)
+  distribution <- match.arg(distribution)
+
+  stopifnot(
+    "outbreak_start_date is only required for adding date of last contact" =
+      contact_type == "last" && !missing(outbreak_start_date) ||
+      contact_type == "first" && missing(outbreak_start_date)
+  )
+
+  rdist <- switch(distribution,
+    pois = stats::rpois,
+    geom = stats::rgeom
+  )
+
+  # c() over ...length() to ensure NULL is not counted by length
+  if (length(c(...)) == 0) {
+    stop("Distribution parameters are missing, check config", call. = FALSE)
+  }
+
+  # name list elements with vec names to ensure arg matching in do.call
+  args <- c(nrow(.data), list(...))
+
+  contact_delay <- tryCatch(
+    error = function(cnd) {
+      stop(
+        "Incorrect parameterisation of distribution, check config",
+        call. = FALSE
+      )
+    },
+    warning = function(cnd) {
+      stop(
+        "Incorrect parameterisation of distribution, check config",
+        call. = FALSE
+      )
+    },
+    do.call(rdist, args = args)
+  )
```

Comment:
Suggestion: Could move this up as the first argument, to be more in line with the function name (try first, catch second). Might make it more readable. 

```suggestion
  contact_delay <- tryCatch(
    do.call(rdist, args = args),
    error = function(cnd) {
      stop(
        "Incorrect parameterisation of distribution, check config",
        call. = FALSE
      )
    },
    warning = function(cnd) {
      stop(
        "Incorrect parameterisation of distribution, check config",
        call. = FALSE
      )
    }
  )
```

Reply:
Resolved in 05b44609f45fb497c5733817f8d0e65e341fd716.